exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/EasyEVFBatchReader.java

```diff
@@ -0,0 +1,329 @@
+ * </ol>
+ *
+ * This class inmplements a series of methods to facilitate schema creation. However, to achieve the best
+ * possible performance, it is vital to use the correct methods to map the schema. Drill will perform fastest when
+ * the schema is known in advance and the writers can be stored in a data structure which minimizes the amount of string
+ * comparisons.
+ *
+ * <b>Current state</b>
+ * For the third scenario, where the schema is not consistent, there are a collection of functions, all named <pre>writeDataType()</pre>
+ * which accept a ScalarWriter, column name and value. These functions first check to see whether the column has been added to the schema, if not
+ * it adds it. If it has been added, the value will be added to the column.
+ *
+ * <p>
+ *
+ *   <pre>
+ *   public boolean open(FileSchemaNegotiator negotiator) {
+ *     super.open(negotiator);
+ *     super.fileIterator = new LTSVRecordIterator(getRowWriter(), reader);
+ *     return true;
+ *   }
+ *   </pre>
+ *
+ */
+public abstract class EasyEVFBatchReader implements ManagedReader<FileSchemaNegotiator> {
+
+  private static final Logger logger = LoggerFactory.getLogger(EasyEVFBatchReader.class);
+
+  public FileSplit split;
+
+  public EasyEVFIterator fileIterator;
+
+  public ResultSetLoader loader;
+
+  private RowSetLoader rowWriter;
+
+  public InputStream fsStream;
+
+  public BufferedReader reader;
+
+  public RowSetLoader getRowWriter() {
+    return rowWriter;
+  }
+
+  @Override
+  public boolean open(FileSchemaNegotiator negotiator) {
+    this.split = negotiator.split();
+    this.loader = negotiator.build();
+    this.rowWriter = loader.writer();
+    try {
+      this.fsStream = negotiator.fileSystem().openPossiblyCompressedStream(split.getPath());
+      this.reader = new BufferedReader(new InputStreamReader(fsStream, StandardCharsets.UTF_8));
+    } catch (IOException e) {
+      throw UserException
+        .dataReadError(e)
+        .message(String.format("Failed to open input file: %s", split.getPath()))
+        .addContext(e.getMessage())
+        .build(logger);
+    }
+    return true;
+  }
+
+  @Override
+  public boolean next() {
+    while (rowWriter.start()) {
+      if (!fileIterator.nextRow()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public void close() {
+    try {
+      AutoCloseables.close(reader, fsStream);
+    } catch (Exception e) {
+      logger.warn("Error closing Input Streams.");
+    }
+  }
+
+  /**
+   * Writes a string column.  This function should be used when the schema is not known in advance or cannot be inferred from the first
+   * row of data.
+   * @param rowWriter The rowWriter object
+   * @param fieldName The name of the field to be written
+   * @param value The value to be written
+   */
+  public static void writeStringColumn(TupleWriter rowWriter, String fieldName, String value) {
```

Comment:
These methods are handy, but they encourage quick-and-dirty (but slow) approaches to reading records. In general, it is best to use indexed mappings from input fields to column writers to avoid a per-value hash.

When I thought about this issue, I realized that most file formats have their own native data types, so conversion is needed from those types to a Java type before we could call the column writer methods. This means we need column-specific "converters". If we have those, then those converters are the right place to hold the corresponding column writer. Else, we could end up with *two* has lookups per value, which would be horrible for performance.

That is, the design we want is:

* For column `i`, obtain the converter for `i`.
* The converter reads data in a source-specific format.
* The converter converts that source-specific format to a Java type.
* The converter holds onto a column writer and writes the Java data to the column writer.

This means each plugin will have its own way to map from source columns to converters: by name (nested, if the source supports maps), or (ideally) by position.

Also, recall that Drill handles arrays and maps which are hard to handle with this approach unless we first materialize the data into a Java array or map, which would be costly.

So, great idea, but the problem is a bit more complex than this approach suggests.

All that said, it would be fine to create a class for the very narrow use case of sources that need no conversion and only work with simple scalar values.